### PR TITLE
Migrate to Dotnet 6

### PR DESCRIPTION
**Pull request by and-rewsmith**
This PR migrates main branch from dotnet 3.1 to dotnet 6.0. This is working off of Damon's prior PR here:
https://github.com/Azure/iotedge/pull/6185

Code changes on top of the linked PR are minor. Only notable things:
1. Removed symlinks from arm64 dockerfiles that are no longer needed
2. TempfilterFunctions module is still on dotnet 3, as base image doesn't support dotnet 6 for arm32
3. EdgeHub watchdog has to shell out to sleep rather than call `thread::sleep()` to bypass a bug in rust. Logged a bug with rust repo.
4. Various retries added in E2E test code to get around problems caused by dotnet 6. Logged bugs with SDK team where appropriate.
5. Upgrade rust artifact build OS for amd64/arm* platforms to ubuntu 20 (from ubuntu 18).

Passing pipelines:
1. [Build Images](https://dev.azure.com/msazure/One/_build/results?buildId=54001600&view=results)
2. [CI](https://dev.azure.com/msazure/One/_build/results?buildId=53619492&view=results)
6. [E2E](https://dev.azure.com/msazure/One/_build/results?buildId=54004817&view=results)
7. [Nested E2E](https://dev.azure.com/msazure/One/_build/results?buildId=53668874&view=results)
8. [Connectivity](https://dev.azure.com/msazure/One/_build/results?buildId=53834451&view=results)
9. [Build Images Release](https://dev.azure.com/msazure/One/_build/results?buildId=53619579&view=results)

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [x] concise summary of tests added/modified
	- [x] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

File: builds/checkin/edgelet.yaml [hunk 1]
```diff
@@ -62,8 +62,8 @@ jobs:
         displayName: Set Version
       - bash: scripts/linux/generic-rust/install.sh --project-root "edgelet"
         displayName: Install Rust
-      - script: "cargo install cross --version ^0.2"
-        displayName: "Install cross"
+      - script: "cargo install --git https://github.com/arsing/cross.git --branch set-path"
+        displayName: "Install cross (fork with docker fix)"
```
Comment (and-rewsmith) [inline builds/checkin/edgelet.yaml:66]: Why do we need Arnav's fork?

File: builds/e2e/e2e-release.yaml [hunk 2]
```diff
@@ -1,180 +0,0 @@
...
-        steps:
-        - task: Docker@2
-          displayName: Docker login edgerelease
-          inputs:
-            command: login
-            containerRegistry: iotedge-release-acr
-        - template: templates/e2e-clean-directory.yaml
-        - template: templates/e2e-setup-base-image-update-release.yaml
-        - template: templates/e2e-clear-docker-cached-images.yaml
-        - template: templates/e2e-run-base-image-update.yaml
```
Comment (and-rewsmith) [inline builds/e2e/e2e-release.yaml:180]: Why is this being removed?

File: builds/e2e/templates/e2e-run-base-image-update.yaml [hunk 3]
```diff
@@ -1,126 +0,0 @@
...
-    Write-Output "##vso[task.setvariable variable=artifactSuffix]$artifactSuffix"
-  displayName: Collect Logs
-  condition: succeededOrFailed()
-
-- task: PublishBuildArtifacts@1
-  displayName: Publish logs
-  inputs:
-    PathtoPublish: $(Build.ArtifactStagingDirectory)/logs${{ parameters.test_type }}
-    ArtifactName: logs-end-to-end-$(artifactSuffix)
-  condition: succeededOrFailed()
```
Comment (and-rewsmith) [inline builds/e2e/templates/e2e-run-base-image-update.yaml:126]: Why is this removed?

File: builds/e2e/e2e.yaml [hunk 4]
```diff
@@ -32,72 +32,22 @@ variables:
...
-    - template: templates/e2e-run.yaml
-
-################################################################################
-  - job: debian_9_arm32v7
-################################################################################
-    displayName: Debian 9 arm32v7 (minimal)
-
-    pool:
-      name: $(pool.custom.name)
-      demands: deb9-e2e-tests
```
Comment (and-rewsmith) [inline builds/e2e/e2e.yaml:92]: Why is this removed?

File: builds/e2e/templates/e2e-setup-base-image-update-release.yaml [hunk 5]
```diff
@@ -1,235 +0,0 @@
...
-  
-      if ($env:AGENT_PROXYURL)
-      {
-        $context['testRunnerProxy'] = $env:AGENT_PROXYURL
-        $context['edgeProxy'] = $env:AGENT_PROXYURL
-      }
-      
-      $context | ConvertTo-Json | Out-File -Encoding Utf8 '$(binDir)/context.json'
-      Get-Content -Path '$(binDir)/context.json'
-    displayName: Create test arguments file (context.json)
```
Comment (and-rewsmith) [inline builds/e2e/templates/e2e-setup-base-image-update-release.yaml:235]: Why is this removed?

File: builds/misc/images-release-base-image-update.yaml [hunk 6]
```diff
@@ -1,575 +0,0 @@
...
-      - task: Docker@2
-        displayName: Docker login edgerelease
-        inputs:
-          command: login
-          containerRegistry: iotedge-release-acr
-      - script: scripts/linux/buildManifest.sh -r $(registry.address) -v $(version) -t $(manifestFilePath) -n '$(namespace)' --tags "$(tags)"
-
-################################################################################
-  - template: ../e2e/e2e-release.yaml
-################################################################################
```
Comment (and-rewsmith) [inline builds/misc/images-release-base-image-update.yaml:575]: Why is this removed?

File: platform-validation/deployment/platform-compatibility.json [hunk 7]
```diff
@@ -1,80 +0,0 @@
...
-            }
-        },
-        "SimulatedTemperatureSensor": {
-            "properties.desired": {
-                "SendData": true,
-                "SendInterval": 5
-            }
-        }
-    }
-}
```
Comment (and-rewsmith) [inline platform-validation/deployment/platform-compatibility.json:80]: Why is this removed?

File: platform-validation/scripts/aziot-compatibility.sh [hunk 8]
```diff
@@ -832,41 +635,24 @@ process_args() {
...
-
-base_message="Azure IoT Compatibility Script had $FAILURES Errors $WARNINGS Warnings and $PASS Successful Checks"
-if [ $FAILURES -gt 0 ]; then
-    wrap_bad "$base_message"
-    exit 1
-elif [ $WARNINGS -gt 0 ]; then
-    wrap_warning_message "$base_message"
-else
-    wrap_pass "$base_message"
 fi
```
Comment (and-rewsmith) [inline platform-validation/scripts/aziot-compatibility.sh:656]: Why is this changed?

File: platform-validation/scripts/check-for-updates.sh [hunk 9]
```diff
@@ -1,325 +0,0 @@
...
-#Compare Usage and Exit if Current Usage exceeds recorded usage
-size_buffer="$(grep "iotedge_size_buffer=" <"$COMPATIBILITY_TOOL_PATH" | sed -r "s/^iotedge_size_buffer=//g")"
-memory_buffer="$(grep "iotedge_memory_buffer=" <"$COMPATIBILITY_TOOL_PATH" | sed -r "s/^iotedge_memory_buffer=//g")"
-cat "$BENCHMARK_OUTPUT_DIR/analyze-memory-logs.out"
-echo "Size buffer is $size_buffer"
-compare_usage container size "$size_buffer"
-compare_usage container memory "$memory_buffer"
-compare_usage binaries size "$size_buffer"
-compare_usage binaries avg_memory "$memory_buffer"
-check_shared_library
```
Comment (and-rewsmith) [inline platform-validation/scripts/check-for-updates.sh:325]: Why is this changed?

File: edge-hub/watchdog/src/child.rs [hunk 10]
```diff
@@ -100,13 +104,24 @@ pub fn run(
...
+
+// We can use `thread::sleep()` instead when this issue is resolved:
+// https://github.com/rust-lang/rust/issues/95661
+fn sleep(duration: Duration) -> Result<()> {
+    Command::new("sleep")
+        .arg(duration.as_secs().to_string())
+        .output()?;
+
+    Ok(())
+}
```
Comment (and-rewsmith) [inline edge-hub/watchdog/src/child.rs:127]: We could only do this for debian 10. Thoughts?

File: test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs [hunk 11]
```diff
@@ -310,7 +312,19 @@ static async Task<LeafDevice> DeleteIdentityIfFailedAsync(Device device, IotHub
...
+            // Can be removed when the below are fixed:
+            // 1 - sometimes tls auth error occurs because EdgeHub sends an unexpected message (work item 14057676)
+            // 2 - devices sdk issue (https://github.com/Azure/azure-iot-sdk-csharp/issues/2337)
+            var retryStrategy = new Incremental(15, RetryStrategy.DefaultRetryInterval, RetryStrategy.DefaultRetryIncrement);
+            var retryPolicy = new RetryPolicy(new FailingTLSConnectionErrorDetectionStrategy(), retryStrategy);
+            await retryPolicy.ExecuteAsync(
+                async () =>
+            {
+                await client.SetMethodHandlerAsync(nameof(DirectMethod), DirectMethod, null, token);
+            }, token);
```
Comment (and-rewsmith) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs:326]: FYA, added this retry

File: test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs [hunk 12]
```diff
@@ -232,6 +234,20 @@ public async Task<Device> CreateEdgeDeviceIdentityAsync(string deviceId, Option<
...
+
+            // Sometimes eventhub will hang when getting runtime information, so we need a retry.
+            var retryStrategy = new Incremental(15, RetryStrategy.DefaultRetryInterval, RetryStrategy.DefaultRetryIncrement);
+            var retryPolicy = new RetryPolicy(new CatchTimeoutErrorDetectionStrategy(), retryStrategy);
+            return await retryPolicy.ExecuteAsync(
+                async () =>
+            {
+                return await Task.Run(async () => await client.GetRuntimeInformationAsync(), eachRequestCancellationToken);
+            }, token);
+        }
```
Comment (and-rewsmith) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs:249]: FYA, added this retry

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh [hunk 13]
```diff
@@ -85,6 +85,8 @@ then
     chmod 600 "$backupjson"
   fi
 
+  echo "$(date --utc +"%Y-%m-%d %H:%M:%S %:z") Completed necessary setup. Starting EdgeAgent."
```
Comment (damonbarry) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh:88]: Is this something we need to check in? If so, it should say "Edge Agent" (with a space), similar to the echo at the [top](https://github.com/Azure/iotedge/blob/2deeffaf3560ed6b6c0cd7afbcca7281fbca94e7/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh#L26) of the script.

**Review by damonbarry — COMMENTED** [review]
Looks good! Just one comment...

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh (hunk 13 above)
Comment (and-rewsmith) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/agentStart.sh:88]: I wanted to check it in because I felt it was confusing for me when an error occurred without this log. I didn't know if the error was coming from a setup step or the dotnet command. Will fix.

**Review by damonbarry — APPROVED** [review]

File: edge-hub/watchdog/src/child.rs (hunk 10 above)
Comment (nimanch) [reply edge-hub/watchdog/src/child.rs:127]: My concern with this is a possible performance penalty. Did we try other versions of sleep, tokio or futures?

File: test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs (hunk 12 above)
Comment (nimanch) [reply test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs:249]: If this does not happen again, My preference would be to remove the retry.

File: edge-hub/watchdog/src/child.rs (hunk 10 above)
Comment (and-rewsmith) [reply edge-hub/watchdog/src/child.rs:127]: We considered introducing tokio runtime but thought it was better to put this in temporarily for now then switch back to thread::sleep when rust bug was fixed. However we recently found that we are removing broker from 1.3 codebase, which means the whole watchdog is disappearing.

File: test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs (hunk 12 above)
Comment (and-rewsmith) [reply test/Microsoft.Azure.Devices.Edge.Test.Common/IotHub.cs:249]: It happens in the pipeline sometimes so I think we need to keep the retry.

File: builds/checkin/dotnet.yaml [hunk 14]
```diff
@@ -11,7 +11,7 @@ jobs:
 ################################################################################
     displayName: Check pipeline preconditions (changes ARE NOT in either edgelet, docs, or mqtt folder)
     pool:
-      vmImage: "ubuntu-18.04"
+      vmImage: "ubuntu-20.04"
```
Comment (nimanch) [inline builds/checkin/dotnet.yaml:14]: does dotnet6 only run on ubuntu 20.04 or this was just an unrelated enhancement?

File: builds/e2e/templates/e2e-setup.yaml [hunk 15]
```diff
@@ -139,7 +139,7 @@ steps:
   - pwsh: |
       $manifestSignerClientDir = '$(Build.SourcesDirectory)/samples/dotnet/ManifestSignerClient'
       dotnet build $manifestSignerClientDir
-      $manifestSignerClientBinDir = Convert-Path "$manifestSignerClientDir/bin/Debug/netcoreapp3.1"
+      $manifestSignerClientBinDir = Convert-Path "$manifestSignerClientDir/bin/Debug/net6.0"
```
Comment (nimanch) [inline builds/e2e/templates/e2e-setup.yaml:142]: You will need to update e2e-setup-base-image-update-release.yaml as well?

File: builds/checkin/dotnet.yaml (hunk 14 above)
Comment (and-rewsmith) [reply builds/checkin/dotnet.yaml:14]: Unrelated enhancement.

File: builds/e2e/templates/e2e-setup.yaml (hunk 15 above)
Comment (and-rewsmith) [reply builds/e2e/templates/e2e-setup.yaml:142]: Yep I think so, good catch. Also compare-compatability.yaml

**Review by nimanch — APPROVED** [review]

**Review by damonbarry — DISMISSED** [review]
